### DISCUSSION TITLE
RAUC Memory Usage

### DISCUSSION THREAD

**Discussion by ghost** [rauc/rauc#1186] [General]
We at delonghi are using am imx8mn-evk with 1gb of RAM and 8gb of flash (emmc)
We have setup rauc to perform A to B updates for just rootfs's
Our rauc bundle size is currently 315mb.

We are currently following the procedure below to perform OTA of RAUC Bundles
1) Download the bundle from a server to the currently running rootfs using curl
2) Perform a rauc install on the bundle

From the above we have noticed the following issues:
1) The bundle is downloaded perfectly fine to the running rootfs
2) RAUC copy and verification process is process works perfectly fine
3) When RAUC does the actual install; it consumes a lot of RAM; at least 650mb
4) W.R.T point 4 above; this cause other running application to stop cause their is not enough RAM
5) Why is RAUC consuming so much RAM?
6) How can RAUC be configured to use less RAM?

From the above please can you tell us why this is happening
Also is their a RAUC configuration to reduce memory usage at the installation stage?

**Comment by ghost** [comment]
I decided to change the bundle configuration to use adaptive updates and this has resolved the issue.
Using streaming or local download then rauc install both seem to work perfectly

**Reply by jluebbe** [reply]
Good to hear. Still, the high memory usage shouldn't happen without adaptive update and I'd like to understand the cause. Please see my other reply: https://github.com/rauc/rauc/discussions/1186#discussioncomment-6499242

**Reply by ghost** [reply]
Hi Jan Lübbe

Using Adaptive Updates consumes very little RAM. So this issue is resolved
Thank you so much for you help and assistance

**Comment by jluebbe** [comment]
> We at delonghi are using am imx8mn-evk with 1gb of RAM and 8gb of flash (emmc) We have setup rauc to perform A to B updates for just rootfs's Our rauc bundle size is currently 315mb.
> 
> We are currently following the procedure below to perform OTA of RAUC Bundles
> 
> 1. Download the bundle from a server to the currently running rootfs using curl
> 2. Perform a rauc install on the bundle

That sounds very normal. I assume you use a ext4 rootfs and have the FS image in the bundle (instead of a tar)?

> From the above we have noticed the following issues:
> 
> 1. The bundle is downloaded perfectly fine to the running rootfs
> 2. RAUC copy and verification process is process works perfectly fine
> 3. When RAUC does the actual install; it consumes a lot of RAM; at leaset 650mb

RAUC itself should should not allocate a lot of memory, as it just copies data from the bundle to the target slot in relatively small chunks. How did you measure the memory usage? I'd suggest:

- `ps faxu` (or something else showing per process RSS, but even better `smemstat`)
- `cat /proc/meminfo`

I suspect you may have a lot of 'Dirty' memory, that sill needs to be written back. The output above should be enough to check if that's the case.

> 4. W.R.T point 4 above; this cause other running application to stop cause their is not enough RAM

> 5. Why is RAUC consuming so much RAM?
> 6. How can RAUC be configured to use less RAM?

If it's dirty data in cache, Linux provides several options to configure write-back: https://www.kernel.org/doc/html/next/admin-guide/sysctl/vm.html

If not, we'd first need to find out what is the root cause of the memory usage.

If you have time to reproduce the issue again, we'd be interested in the additional information requested above.

**Reply by ghost** [reply]
We are using ex4 instead of tar
We measured memory using the linux free command polling it every 1 second via ssh
We are clearing the ram caches using vm.drop_caches

We can re-produce the issue 100% on every rauc install command

**Reply by jluebbe** [reply]
> We are using ex4 instead of tar
> We measured memory using the linux free command polling it every 1 second via ssh

The 'free' command doesn't provide enough details to analyse issues such as this. Please add the output of `cat /proc/meminfo` while the memory usage is high.

> We are clearing the ram caches using vm.drop_caches

OK. I assume that it doesn't help? That would be consistent with a high amount of dirty pages, as these can't be dropped via vm.drop_caches.

To clean up dirty pages, you can run `sync`. If that takes some time, it's a strong indication that we're dealing with many dirty pages here.

> We can re-produce the issue 100% on every rauc install command

Good, so we can hopefully find the root cause. :)

**Reply by ghost** [reply]
i will use cat /proc/meminfo instead of free
also i will use sync aswell as vm.drop_caches. to clean up dirty pages

will keep you posted with the above changes
:-)

**Reply by ghost** [reply]
please see below a log of cat /proc/meminfo

```
MemTotal:         962724 kB
MemFree:           88436 kB
MemAvailable:     632812 kB
Buffers:          272624 kB
Cached:           372568 kB
SwapCached:            0 kB
Active:            16908 kB
Inactive:         704380 kB
Active(anon):        164 kB
Inactive(anon):   151244 kB
Active(file):      16744 kB
Inactive(file):   553136 kB
Unevictable:           0 kB
Mlocked:               0 kB
SwapTotal:             0 kB
SwapFree:              0 kB
Dirty:            154736 kB
Writeback:          3388 kB
AnonPages:         76108 kB
Mapped:            57468 kB
Shmem:             75300 kB
KReclaimable:      64608 kB
Slab:              94980 kB
SReclaimable:      64608 kB
SUnreclaim:        30372 kB
KernelStack:        2928 kB
PageTables:         2812 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:      481360 kB
Committed_AS:     511928 kB
VmallocTotal:   135290159040 kB
VmallocUsed:        8864 kB
VmallocChunk:          0 kB
Percpu:              704 kB
HardwareCorrupted:     0 kB
AnonHugePages:     26624 kB
ShmemHugePages:        0 kB
ShmemPmdMapped:        0 kB
FileHugePages:         0 kB
FilePmdMapped:         0 kB
CmaTotal:          65536 kB
CmaFree:           34804 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:               0 kB
```

**Reply by ghost** [reply]
look at 
`Dirty:            154736 kB`
is this too large?

**Reply by jluebbe** [reply]
~155 MB of dirty pages is plausible during image writing. But, compared to `MemAvailable:     632812 kB`, that's not very much. `MemAvailable` includes `Cached:           372568 kB`, which would be reclaimed by the kernel as soon as another application requests memory. So this doesn't look like it would trigger the kernel's out-of-memory killer.

While writeback of dirty pages from RAUC is ongoing, other applications writing to the same storage may also see lower performance.

You said "this cause other running application to stop cause their is not enough RAM". Do you mean that they crash? Or are the pausing or slower than normal?

**Reply by ghost** [reply]
"this cause other running application to stop cause their is not enough RAM". --> the kernel seg faults 1 in 5 times. some applications just dont have enough RAM to keep running and stop. These applications run form systemd services

**Reply by jluebbe** [reply]
Could you post the full kernel out-of-memory message? It would give more details why exactly it couldn't free more memory.

**Comment by ghost** [comment]
Hi Jan Lübbe

I have modified out RAUC configuration to use Adaptive Updates and this has
resolved the problem
Many thanks for your help

Regards

Pritpal

On Thu, 20 Jul 2023 at 13:53, Jan Lübbe ***@***.***> wrote:

> Good to hear. Still, the high memory usage shouldn't happen without
> adaptive update and I'd like to understand the cause. Please see my other
> reply: #1186 (comment)
> <https://github.com/rauc/rauc/discussions/1186#discussioncomment-6499242>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rauc/rauc/discussions/1186#discussioncomment-6499258>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANVR26AIQEGJAIREVLWSEW3XRES5BANCNFSM6AAAAAA2P5IEOU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

-- 
This e-mail is intended only for the person to whom it is addressed and may 
contain confidential information. If an addressing or transmission error 
has misdirected this e-mail, please notify the author by replying to this 
e-mail. If you are not the intended recipient you must not use, disclose, 
print or rely on the contents of this e-mail and should immediately and 
permanently delete it.


Kenwood Limited is a company registered in England 
with registered number 872044. The registered office of Kenwood Limited is 
1 Kenwood Business Park, New Lane, Havant, Hampshire, PO9 2NH.


Kenwood 
Limited may monitor outgoing and incoming e-mails and other 
telecommunications on its e-mail and telecommunications systems for the 
purpose of ensuring compliance with the law and with Kenwood Limited's 
policies.

**Comment by ghost** [comment]
Hi
Will do tomorrow morning

Br

P

On Thu, 20 Jul 2023 at 17:10, Jan Lübbe ***@***.***> wrote:

> Could you post the full kernel out-of-memory message? It would give more
> details why exactly it couldn't free more memory.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rauc/rauc/discussions/1186#discussioncomment-6501354>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANVR26FW34FHI5Q4C4MKCUDXRFJ5VANCNFSM6AAAAAA2P5IEOU>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

-- 
This e-mail is intended only for the person to whom it is addressed and may 
contain confidential information. If an addressing or transmission error 
has misdirected this e-mail, please notify the author by replying to this 
e-mail. If you are not the intended recipient you must not use, disclose, 
print or rely on the contents of this e-mail and should immediately and 
permanently delete it.


Kenwood Limited is a company registered in England 
with registered number 872044. The registered office of Kenwood Limited is 
1 Kenwood Business Park, New Lane, Havant, Hampshire, PO9 2NH.


Kenwood 
Limited may monitor outgoing and incoming e-mails and other 
telecommunications on its e-mail and telecommunications systems for the 
purpose of ensuring compliance with the law and with Kenwood Limited's 
policies.

**Reply by ghost** [reply]
I personally think tha problem is that we only have 1gb or RAM and we are running a number of applications that reduces free RAM to 650mb; the compressed rauc bundle is 300mb.

So basically we just don't have enough RAM to perform a rauc install.
1) Is their a way to reduce the bundle size?

**Reply by ghost** [reply]
Regarding a log of the crash; thats is hard to reproduce cause the rauc install takes place; finishes. But other applications stop responding. The kernel does not crash; but only a sometimes a couple of device drivers like rpmsg get trashed and the CMA gets over used

**Reply by jluebbe** [reply]
> I personally think tha problem is that we only have 1gb or RAM and we are running a number of applications that reduces free RAM to 650mb; the compressed rauc bundle is 300mb.
> 
> So basically we just don't have enough RAM to perform a rauc install.

You mentioned that you store the bundle on the running rootfs, so at no point in time the full bundle needs to be stored fully in RAM. RAUC should only use a few MBs itself + read&write caches. (which are managed by the kernel)

> 1. Is their a way to reduce the bundle size?

You might increase compression, but that is unlikely to help you here, as the caches store uncompressed data.

**Reply by jluebbe** [reply]
> Regarding a log of the crash; thats is hard to reproduce cause the rauc install takes place; finishes. But other applications stop responding. The kernel does not crash; but only a sometimes a couple of device drivers like rpmsg get trashed and the CMA gets over used

Problems with CMA or rpmsg should always generate error messages in the kernel log. Without those exact messages it's impossible to be sure what is actually happening here. Any error messages should be shown on the serial console and should be easy to collect.

**Reply by ghost** [reply]
Regarding rpmsg we get errors that the virtio queue is out of range. When I see this message any others I will upload them.
What I have noticed is the following:
1) rauc install - at 90% copy bundle to rootfs.1 partition 
   This is fine and the free memory is about 650mb
2)  rauc install - at installation stage; our free memory drops to  2gb; and other aps stop.
     So why at the last stage of the rauc installation does rauc consume so much RAM?
     Is their a configuration that we have missed?

**Reply by jluebbe** [reply]
If you see copy bundle at 90%, you're probably using an old release of RAUC. Please update to v1.10, which will give you better progress information.

How do you see 2GB "free" with 1GB RAM?

"copy bundle" is when RAUC reads from the bundle and writes to the slot, which will use memory for the kernel's write cache. Have you tried the write-back configuration options I referred to in https://github.com/rauc/rauc/discussions/1186#discussioncomment-6499242 (i.e. setting `dirty_ratio` to a much lower value)?

**Reply by ghost** [reply]
How do you see 2GB "free" with 1GB RAM?
Sorry this is a typo i meant at installation stage; our free memory drops to 2mb;
Also we are using rauc version 1.9; is their a newer version?

**Reply by ghost** [reply]
regardinf dirty_ratio we are now using sync aswell; hence every 1 second we issue the sync command

**Reply by jluebbe** [reply]
The releases are listed at https://github.com/rauc/rauc/releases. Does calling sync regularly make a difference?

**Comment by ghost** [comment]
calling sync regularly make a difference?
has no effect

**Comment by ghost** [comment]
Please see below the kernel crash log that we get during the rauc install

```
[ 5205.227978] Unable to handle kernel paging request at virtual address 0000001e00000428
[ 5205.235962] Mem abort info:
[ 5205.239125]   ESR = 0x96000004
[ 5205.242223]   EC = 0x25: DABT (current EL), IL = 32 bits
[ 5205.247618]   SET = 0, FnV = 0
[ 5205.250753]   EA = 0, S1PTW = 0
[ 5205.253934] Data abort info:
[ 5205.257087]   ISV = 0, ISS = 0x00000004
[ 5205.260972]   CM = 0, WnR = 0
[ 5205.264012] user pgtable: 4k pages, 48-bit VAs, pgdp=00000000481ff000
[ 5205.270484] [0000001e00000428] pgd=0000000048df0003, p4d=0000000048df0003, pud=0000000000000000
[ 5205.279230] Internal error: Oops: 96000004 [#1] PREEMPT SMP
[ 5205.285145] Modules linked in: esp32_sdio(O) fsl_jr_uio caam_jr caamkeyblob_desc caamhash_desc caamalg_desc crypto_engine rng_core libdes crct10dif_ce caam secvio error imx_y
[ 5205.301435] CPU: 1 PID: 103023 Comm: kworker/1:2H Tainted: G           O      5.10.9-1.0.0+g3cd8d91edbc1 #1
[ 5205.311173] Hardware name: NXP i.MX8MNano EVK board (DT)
[ 5205.316493] Workqueue: kblockd blk_mq_run_work_fn
[ 5205.321203] pstate: 60000005 (nZCv daif -PAN -UAO -TCO BTYPE=--)
[ 5205.327211] pc : __blk_rq_map_sg+0x160/0x58c
[ 5205.331481] lr : mmc_queue_map_sg+0x3c/0x60
[ 5205.335661] sp : ffff800017d5b9d0
[ 5205.338973] x29: ffff800017d5b9d0 x28: 0000000000000001 
[ 5205.344286] x27: 0000000000000800 x26: ffff000038420078 
[ 5205.349598] x25: 00000000685be800 x24: 0000000000000200 
[ 5205.354910] x23: 0000000000000000 x22: 0000000000000001 
[ 5205.360222] x21: 0000001e00000400 x20: ffff0000055923c0 
[ 5205.365533] x19: ffff800017d5bad0 x18: 0000000000000000 
[ 5205.370846] x17: 0000000000000000 x16: 0000000000000000 
[ 5205.376157] x15: 0000003410f0b22d x14: 00000000000001df 
[ 5205.381469] x13: 0000000000000001 x12: ffff800011719000 
[ 5205.386781] x11: 0000000000000000 x10: 0000000000000600 
[ 5205.392093] x9 : 0000000000000000 x8 : ffff000005592640 
[ 5205.397405] x7 : fffffdffffe00000 x6 : ffff000004e03dc0 
[ 5205.402716] x5 : 0000000000000200 x4 : 0000000000000001 
[ 5205.408028] x3 : ffff000038420078 x2 : ffff0000055ac000 
[ 5205.413339] x1 : fffffe0000816f80 x0 : 0000000000000200 
[ 5205.418652] Call trace:
[ 5205.421101]  __blk_rq_map_sg+0x160/0x58c
[ 5205.425022]  mmc_queue_map_sg+0x3c/0x60
[ 5205.428858]  mmc_blk_data_prep+0x1ac/0x34c
[ 5205.432953]  mmc_blk_mq_issue_rq+0x498/0x870
[ 5205.437222]  mmc_mq_queue_rq+0x120/0x2bc
[ 5205.441143]  blk_mq_dispatch_rq_list+0x104/0x760
[ 5205.445760]  __blk_mq_sched_dispatch_requests+0xd0/0x1c0
[ 5205.451071]  blk_mq_sched_dispatch_requests+0x3c/0x80
[ 5205.456120]  __blk_mq_run_hw_queue+0xa0/0x10c
[ 5205.460474]  blk_mq_run_work_fn+0x24/0x30
[ 5205.464485]  process_one_work+0x1cc/0x350
[ 5205.468493]  worker_thread+0x13c/0x470
[ 5205.472241]  kthread+0x154/0x160
[ 5205.475468]  ret_from_fork+0x10/0x30
[ 5205.479047] Code: f90067e1 291a6ff8 f94002b5 b4001e55 (294572a9) 
[ 5205.485140] ---[ end trace da2793d2ec85fd15 ]---
```

Broadcast message from systemd-journald@imx8mnevk (Fri 2023-07-21 08:51:06 UTC):

kernel[292]: [ 5205.279230] Internal error: Oops: 96000004 [#1] PREEMPT SMP


Broadcast message from systemd-journald@imx8mnevk (Fri 2023-07-21 08:51:06 UTC):

kernel[292]: [ 5205.479047] Code: f90067e1 291a6ff8 f94002b5 b4001e55 (294572a9)

**Comment by ghost** [comment]
Also using the journalctl -u rauc why do we get this error and how do we resolve it

    Jul 21 08:49:43 imx8mnevk rauc[111101]: [[0;1;38;5;185m[[0;1;39m[[0;1;38;5;185mContinuing after adaptive mode error: failed to open target slot hash index for rootfs.1: data file size (1621870592) is not a multiple of 4096 bytes[[0m
